core/trino-main/src/main/java/io/trino/execution/QueryManagerConfig.java

```diff
@@ -77,6 +77,7 @@ public class QueryManagerConfig
     private int queryManagerExecutorPoolSize = 5;
     private int queryExecutorPoolSize = 1000;
     private int maxStateMachineCallbackThreads = 5;
+    private int maxSplitManagerCallbackThreads = 100;
```

Comment:
Do these threads do actual work or they just process events?

Reply:
These threads call `ConnectorSplitSource#getNextBatch`, so they will do whatever work those methods perform before returning a future.
Potentially the number of threads calling BufferingSplitSource is the number of table scans which are being scheduled in the cluster. I don't think there is an upper limit other than `query.executor-pool-size` or `query.max-concurrent-queries` today.

Reply:
100 feels like a lot. But maybe does not matter much.

Reply:
Do all major connectors return future rather than enumerating splits directly?

Reply:
Among the data lake connectors, iceberg is the only one enumerating directly and then returning a completed future. Most connectors seem to use FixedSplitSource where there is not much work done in getNextBatch.

> 100 feels like a lot

I wasn't too sure how much is enough here. `query.executor-pool-size` defaults to 1000, so I tried to choose some number which won't make this thread pool the limiting factor in most cases.

Reply:
> Among the data lake connectors, iceberg is the only one enumerating directly and then returning a completed future. Most connectors seem to use FixedSplitSource where there is not much work done in getNextBatch.

@findepi that's probably something to fix. Is iceberg enumeration IO heavy or CPU heavy. If the former, then I could see the case where 100 might not be enough

Reply:
> Among the data lake connectors, iceberg is the only one enumerating directly and then returning a completed future. Most connectors seem to use FixedSplitSource

there is also BackgroundHiveSplitLoader for Hive

for Delta currently most work is done when constructing split source, but that's going to change. we gonna process the delta checkpoint lazily

Reply:
> there is also BackgroundHiveSplitLoader for Hive

This is async as there is a split queue.

Reply:
@raunaqmorarka maybe we make the value as small as number of calling threads?

Reply:
How would we derive value of number of calling threads ? As far as I can tell there isn't a static upper bound on the number of table scans which are being scheduled in the cluster.

Reply:
> As far as I can tell there isn't a static upper bound on the number of table scans which are being scheduled in the cluster.

I though there is some bounded executor, but maybe there isn't? should we make this executor unbounded too?

Reply:
There is `query.executor-pool-size` which defaults to 1000, so there is a bound but it's quite large and not specifically for splits scheduling. Making the executor unbounded might create a lot of threads.

Reply:
Ok. I'm leaving it to your discretion. I would maybe set it to 500.

Reply:
I ran TPC and concurrency benchmarks (32 concurrency) and didn't see adverse impact. Also for connectors which return completed futures we no longer even use the executor pool. So I'm going to keep the current value and revise later if needed.

Reply:
>  ran TPC and concurrency benchmarks

Was it for delta?

Reply:
I ran TPC on hive and iceberg and concurrency on hive

Reply:
I think @findepi mentioned that delta uses these threads for building splits, but maybe it's not that big of a problem